conjure/objectwriter.go

```diff
@@ -93,3 +129,85 @@ func writeStructMarshalInitDecls(methodBody *jen.Group, fields []*types.Field, r
 		}
 	}
 }
+
+func getSafetyFromType(fieldType types.Type, safetyCache map[types.Type]spec.LogSafety) spec.LogSafety {
+	if safetyCache == nil {
+		safetyCache = make(map[types.Type]spec.LogSafety)
+	}
+
+	// Check cache first
+	if cached, exists := safetyCache[fieldType]; exists {
+		return cached
+	}
+
+	// Insert placeholder to detect cycles - use UNKNOWN as safe default
+	safetyCache[fieldType] = spec.New_LogSafety(spec.LogSafety_UNKNOWN)
+
+	// First check if the field type itself has safety annotations
+	fieldSafety := fieldType.Safety()
+	if !fieldSafety.IsUnknown() {
+		// Update cache with real result
+		safetyCache[fieldType] = fieldSafety
+		return fieldSafety
+	}
+
+	// If it's an ObjectType, recursively compute safety from its fields
+	if objectType, ok := fieldType.(*types.ObjectType); ok {
+		result := computeStructSafety(objectType.Fields, safetyCache)
+		// Update cache with computed result
+		safetyCache[fieldType] = result
+		return result
+	}
+
+	// For other types without explicit safety, return unknown so no tags are added
+	// Cache already contains UNKNOWN, so just return it
+	return spec.New_LogSafety(spec.LogSafety_UNKNOWN)
+}
+
+// getSafetyFromField returns the safety of a field, considering both field-level and type-level safety annotations
+func getSafetyFromField(field *types.Field, safetyCache map[types.Type]spec.LogSafety) spec.LogSafety {
+	// First check if the field has an explicit safety annotation
+	if field.Safety != nil {
+		return *field.Safety
+	}
+	// Fall back to type-based safety
+	return getSafetyFromType(field.Type, safetyCache)
+}
+
+func computeStructSafety(fields []*types.Field, safetyCache map[types.Type]spec.LogSafety) spec.LogSafety {
```

Comment:
Sorry, missed this in the first review.

Let's be more precise with the language/naming and parameter -- this function is computing the safety for a particular Conjure Object type, so I think the following naming for function and parameter is clearer:

```suggestion
func computeObjectSafety(object *types.ObjectType, typeSafetyCache map[types.Type]spec.LogSafety) spec.LogSafety {
```

`struct` is not a Conjure concept -- the `conjure-go` generator does generate `struct` code for Conjure objects, so for functions dealing with the generated struct code it makes sense to have `struct` as part of naming, but here it should be `object`

Reply:
Done.